book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -0,0 +1,697 @@
+    blocks: BTreeMap<block::Height, Arc<Block>>,
+    height_by_hash: HashMap<block::Hash, block::Height>,
+    tx_by_hash: HashMap<transaction::Hash, (block::Height, tx_index)>,
+
+    utxos: HashSet<transparent::Output>,
+    sapling_anchors: HashSet<sapling::tree::Root>,
+    sprout_anchors: HashSet<sprout::tree::Root>,
+    sapling_nullifiers: HashSet<sapling::Nullifier>,
+    sprout_nullifiers: HashSet<sprout::Nullifier>,
+    partial_cumulative_work: PartialCumulativeWork,
+}
+```
+
+#### `pub fn push(&mut self, block: Arc<Block>) -> Result<(), Error>`
+
+Push a block into a chain as the new tip if the block is a valid extension of
+that chain.
+
+1. Run contextual validation checks on block against Self
+
+1. Update cumulative data members
+    - Add block to end of `self.blocks`
+    - Add hash to `height_by_hash`
+    - for each `transaction` in `block`
+      - add key: `transaction.hash` and value: `(height, tx_index)` to `tx_by_hash`
+    - Add new utxos and remove consumed utxos from `self.utxos`
+    - Add anchors to the appropriate `self.<version>_anchors`
+    - Add nullifiers to the appropriate `self.<version>_nullifiers`
+    - Add work to `self.partial_cumulative_work`
+
+#### `pub fn pop_root(&mut self) -> Arc<Block>`
+
+Remove the lowest height block of the non-finalized portion of a chain.
+
+1. Remove the lowest height block from `self.blocks`
+
+1. Update cumulative data members
+    - Remove the block's hash from `self.height_by_hash`
+    - for each `transaction` in `block`
+      - remove `transaction.hash` from `tx_by_hash`
+    - Remove new utxos from `self.utxos`
+    - Remove the anchors from the appropriate `self.<version>_anchors`
+    - Remove the nullifiers from the appropriate `self.<version>_nullifiers`
+
+1. Return the block
+
+**Note**: We do not subtract work from `self.partial_cummulative_work`. This
+is to make sure that the ordering of chains is stable while finalizing blocks.
+
+#### `pub fn fork(&self, new_tip: block::Hash) -> Option<Self>`
+
+Fork a chain at the block with the given hash, if it is part of this chain.
+
+1. If `self` does not contain `new_tip` return `None`
+
+2. Clone self as `forked`
+
+3. While the tip of `forked` is not equal to `new_tip`
+   - call `forked.pop_tip()` and discard the old tip
+
+4. Return `forked`
+
+#### `fn pop_tip(&mut self) -> Arc<Block>`
+
+Remove the highest height block of the non-finalized portion of a chain.
+
+1. Remove the highest height `block` from `self.blocks`
+
+2. Update cumulative data members
+    - Remove the corresponding hash from `self.height_by_hash`
+    - for each `transaction` in `block`
+      - remove `transaction.hash` from `tx_by_hash`
+    - Add consumed utxos and remove new utxos from `self.utxos`
+    - Remove anchors from the appropriate `self.<version>_anchors`
+    - Remove the nullifiers from the appropriate `self.<version>_nullifiers`
+    - Subtract work from `self.partial_cumulative_work`
+
+3. Return the block
+
```

Comment:
**Note:** The genesis block is never popped from the tip, so there is no need to special-case it here.

Reply:
also superceded like the earlier genesis block comments I think